base_layer/core/src/base_node/chain_metadata_service/service.rs

```diff
@@ -0,0 +1,394 @@
+        Self {
+            liveness,
+            base_node,
+            peer_chain_metadata: Vec::new(),
+            last_chainstate_flushed_at: Utc::now().naive_utc(),
+        }
+    }
+
+    /// Run the service
+    pub async fn run(mut self) {
+        let mut liveness_event_stream = self.liveness.get_event_stream_fused();
+        let mut base_node_event_stream = self.base_node.get_block_event_stream_fused();
+
+        log_if_error!(
+            target: LOG_TARGET,
+            "Error when updating liveness chain metadata: '{}'",
+            self.update_liveness_chain_metadata().await
+        );
+
+        loop {
+            futures::select! {
+                event = base_node_event_stream.select_next_some() => {
+                    log_if_error!(
+                        level: debug,
+                        target: LOG_TARGET,
+                        "Failed to handle base node event because '{}'",
+                        self.handle_block_event(&event).await
+                    );
+                },
+
+                liveness_event = liveness_event_stream.select_next_some() => {
+                    log_if_error!(
+                        target: LOG_TARGET,
+                        "Failed to handle liveness event because '{}'",
+                        self.handle_liveness_event(&liveness_event).await
+                    );
+                },
+
+                complete => {
+                    info!(target: LOG_TARGET, "ChainStateSyncService is exiting because all tasks have completed");
+                    break;
+                }
+            }
+        }
+    }
+
+    /// Handle BlockEvents
+    async fn handle_block_event(&mut self, event: &BlockEvent) -> Result<(), ChainMetadataSyncError> {
+        match event {
+            BlockEvent::Verified((_, BlockAddResult::Ok(_))) => {
+                self.update_liveness_chain_metadata().await?;
+            },
+            BlockEvent::Verified(_) | BlockEvent::Invalid(_) => {},
+        }
+
+        Ok(())
+    }
+
+    /// Send this node's metadata to
+    async fn update_liveness_chain_metadata(&mut self) -> Result<(), ChainMetadataSyncError> {
+        let chain_metadata = self.base_node.get_metadata().await?;
+        let bytes = proto::ChainMetadata::from(chain_metadata).to_encoded_bytes()?;
+        self.liveness
+            .set_pong_metadata_entry(MetadataKey::ChainMetadata, bytes)
+            .await?;
+        Ok(())
+    }
+
+    async fn handle_liveness_event(&mut self, event: &LivenessEvent) -> Result<(), ChainMetadataSyncError> {
+        match event {
+            // Received a pong, check if our neighbour sent it and it contains ChainMetadata
+            LivenessEvent::ReceivedPong(event) => {
+                if event.is_neighbour {
+                    self.collect_chain_state_from_pong(&event.node_id, &event.metadata)?;
+
+                    // All peers have responded in this round, send the chain metadata to the base node service
```

Comment:
I'm not sure this should go to the Base Node Service, I can’t currently see the benefit. It could pass through the Base Node service, but it will be simpler just to publish the chainmetadata on the chainmetadata event stream and allowing other service to get access to this event stream using the handle as is currently implemented.

Reply:
Agreed, might as well hook that up in this PR

Reply:
Done :)